cmd/zoekt-sourcegraph-indexserver/main.go

```diff
@@ -1049,6 +1076,177 @@ func (s *Server) forceIndex(id uint32) (string, error) {
 	return fmt.Sprintf("Indexed %s with state %s", args.String(), state), nil
 }
 
+// Index implements the gRPC method for indexing a repository. If the repository
+// exists in the trash, it will attempt to recover it before indexing. The
+// method returns metadata about the indexed repository including branches and
+// and the time of indexing.  The metadata might be empty if we fail to read the
+// metadata after the index or if the repository is empty.
+//
+// Possible GRPC error codes:
+// - InvalidArgument: returned when the index options are missing or invalid
+// - AlreadyExists: returned when the repository is already being indexed by another process
+// - Internal: returned when indexing fails
+// - Canceled: returned when the context is canceled while waiting for an index slot
+func (s *Server) Index(ctx context.Context, req *indexserverv1.IndexRequest) (*indexserverv1.IndexResponse, error) {
+	return s.indexGRPC(ctx, req, s.index)
+}
+
+// indexGRPC takes an index func which is used for testing. For production GRPC calls, use
+// Server.Index instead
+func (s *Server) indexGRPC(ctx context.Context, req *indexserverv1.IndexRequest, indexFunc func(*indexArgs) (indexState, error)) (*indexserverv1.IndexResponse, error) {
+	// Validate request
+	if req.Options == nil {
+		return nil, status.Error(codes.InvalidArgument, "index options are required")
+	}
+
+	// Try to acquire semaphore, but respect context cancellation
+	select {
+	case s.indexSemaphore <- struct{}{}:
+		defer func() { <-s.indexSemaphore }()
+	case <-ctx.Done():
+		return nil, status.Error(codes.Canceled, "context canceled while waiting for index slot")
+	}
+
+	var opts IndexOptions
+	opts.FromProto(req.Options)
+	opts.CloneURL = s.rootURL.ResolveReference(&url.URL{Path: path.Join("/.internal/git", req.Options.Name)}).String()
+	args := s.indexArgs(opts)
+
+	// Recover the index from the trash if possible. First we have to make sure
+	// the shard doesn't exist in the index dir, because it might potentially be
+	// overwritten if we recover a shard from the trash. If we can find a shard
+	// in the index dir, we skip recovery. In both cases we continue with index
+	// and let the index process figure out whether the shard is up-to-date or
+	// requires a reindex.
+	if _, _, ok, err := args.BuildOptions().FindRepositoryMetadata(); err == nil && !ok {
+		// The shard is not in the index dir. Now check the trash.
+		trashShards := getShards(filepath.Join(s.IndexDir, ".trash"))
+		if shards, ok := trashShards[opts.RepoID]; ok {
+			infoLog.Printf("restoring shards from trash for %d", opts.RepoID)
+			moveAll(s.IndexDir, shards)
+		}
+	}
+
+	// Index the repository.
+	var indexErr error
+	var indexTimeUnix int64
+	var zoektRepo *zoekt.Repository
+
+	ran := s.muIndexDir.With(opts.Name, func() {
```

Comment:
I think the request context should be propagated down to `gitIndex`, otherwise the indexing process won't respect the client's context deadline. Same thought for deletion.